cranelift/fuzzgen/src/bin/check_opcode_signatures.rs

```diff
@@ -0,0 +1,138 @@
+use cranelift::codegen::ir::{instructions::ResolvedConstraint, types::*, Opcode, Type};
+use cranelift_fuzzgen::OPCODE_SIGNATURES;
+use std::collections::HashMap;
+
+/// The list of opcodes that we exclude from the analysis. This should only be instructions that
+/// affect control flow, as everything else should be handled by the general machinery in
+/// function_generator.
+const SKIPPED_OPCODES: &[Opcode] = &[
+    // These opcodes have special handling in cranelift-fuzzgen
+    Opcode::Call,
+    Opcode::Return,
+    Opcode::Jump,
+    Opcode::Brif,
+    Opcode::BrTable,
+    // TODO: ExtractVector produces dynamic vectors, and those cause a panic in
+    // `constraint.result_type`.
+    Opcode::ExtractVector,
+];
+
+/// This is the set of types that we know how to fuzz in cranelift. It's not specialized by
+/// targets, as we expect any target-specific support for things like SIMD to be expressed in the
+/// `function_generator::valid_for_target` predicate instead.
+const TYPES: &[Type] = &[
+    I8, I16, I32, I64, I128, // Scalar Integers
+    F32, F64, // Scalar Floats
+    I8X16, I16X8, I32X4, I64X2, // SIMD Integers
+    F32X4, F64X2, // SIMD Floats
+];
+
+fn instruction_instantiations<'a>() -> Vec<(Opcode, Vec<Type>, Vec<Type>)> {
+    let mut insts = vec![];
+
+    for op in Opcode::all() {
+        if SKIPPED_OPCODES.contains(op) {
+            continue;
+        }
+
+        let constraints = op.constraints();
+
+        let ctrl_types = if let Some(ctrls) = constraints.ctrl_typeset() {
+            Vec::from_iter(TYPES.iter().copied().filter(|ty| ctrls.contains(*ty)))
+        } else {
+            vec![INVALID]
+        };
+
+        for ctrl_type in ctrl_types {
+            let rets = Vec::from_iter(
+                (0..constraints.num_fixed_results()).map(|i| constraints.result_type(i, ctrl_type)),
+            );
+
+            let mut cols = vec![];
+
+            for i in 0..constraints.num_fixed_value_arguments() {
+                match constraints.value_argument_constraint(i, ctrl_type) {
+                    ResolvedConstraint::Bound(ty) => cols.push(Vec::from([ty])),
+                    ResolvedConstraint::Free(tys) => cols.push(Vec::from_iter(
+                        TYPES.iter().copied().filter(|ty| tys.contains(*ty)),
+                    )),
+                }
+            }
+
+            let mut argss = vec![vec![]];
+
+            let mut cols = cols.as_slice();
+            while let Some((col, rest)) = cols.split_last() {
+                cols = rest;
+
+                let mut next = vec![];
+                for current in argss.iter() {
+                    for ty in col {
+                        let mut args = vec![*ty];
+                        args.extend_from_slice(&current);
+                        next.push(args);
+                    }
+                }
+
+                let _ = std::mem::replace(&mut argss, next);
+            }
```

Comment:
This took me a while to figure out, but I think its the same thing I wrote in my [proof of concept code](https://github.com/afonso360/cranelift-opcode-gen/blob/75bb630898fef4761d8058dd5a5582364e301ca4/src/main.rs#L15-L37) a while ago!

Adding some sort of comment explaining this operation at a high level would be really helpful. 

In my POC I borrowed the name from [Itertools::multi_cartesian_product](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.multi_cartesian_product) which does the same operation, I'm not sure if that is the correct mathematical name for the operation, but it made sense at the time.

Reply:
That's definitely the right name -- I'll add some comments so this thing isn't quite so opaque :)